File: rust/src/layer.rs
use crate::activation;
use crate::error;
use crate::matrix::Matrix;

pub struct ForwardPassResult {
    pub sum: Matrix,
    pub activated: Matrix,
    pub error: f64,
}

pub struct Layer {
    pub is_output: bool,
    pub neuron_count: u32,
    pub weights: Matrix,
    pub weights_transpose: Matrix,
    activation_key: activation::ActivationFunctionKey,
    activation: activation::ActivationFN,
    error_key: error::ErrorFunctionKey,
    error: error::ErrorFN,
}

impl Layer {
    pub fn forward_pass(&self, input: &Matrix, expected: &Matrix) -> ForwardPassResult {
        let sum = self.weights.multiply(input);

        let activated = (self.activation.forward)(&sum, self.is_output);

        let error = if self.is_output {
            (self.error.loss)(&activated, expected)
        } else {
            -1.0
        };
        ForwardPassResult {
            sum,
            activated,
            error,
        }
    }

    pub fn backward_pass(
        &self,
        result: &ForwardPassResult,
        delta_next: &Matrix,
        layer_next: &Layer,
    ) -> Matrix {
        let weights = &layer_next.weights_transpose;
        let backward = self.activation.backward;
        weights
            .multiply(delta_next)
            .hadamard(&backward(&result.sum))
    }

    pub fn output_pass(&self, result: &ForwardPassResult, expected: &Matrix) -> Matrix {
        if let activation::ActivationFunctionKey::Softmax = self.activation_key {
            if let error::ErrorFunctionKey::CrossEntropy = self.error_key {
                return (self.activation.output)(&result.activated, expected);
            } else {
                panic!("can only use softmax with cross-entropy");
            }
        }
        let error = (self.error.grad)(&result.activated, expected);
        let backward = (self.activation.backward)(&result.sum);

        error.hadamard(&backward)
    }

    pub fn set_weights(&mut self, weights: Matrix) {
        self.weights = weights;
        self.weights_transpose = self.weights.omit(0).transpose();
    }

    pub fn initialize(&mut self, weights: Matrix, is_output: bool) {
        self.set_weights(weights);
        self.is_output = is_output;
    }
}

pub fn create_layer(
    count: u32,
    activation: activation::ActivationFunctionKey,
    error: error::ErrorFunctionKey,
) -> Layer {
    let error_fn = match error {
        error::ErrorFunctionKey::CrossEntropy => error::CE_LOSS,
        error::ErrorFunctionKey::LogLoss => error::LOG_LOSS,
        error::ErrorFunctionKey::MeanSquared => error::MSE_LOSS,
    };
    let activation_fn = match activation {
        activation::ActivationFunctionKey::ReLu => activation::RELU_ACTIVATION,
        activation::ActivationFunctionKey::Sigmoid => activation::SIGMOID_ACTIVATION,
        activation::ActivationFunctionKey::TanH => activation::TANH_ACTIVATION,
        activation::ActivationFunctionKey::Softmax => activation::SOFTMAX_ACTIVATION,
    };
    Layer {
        is_output: false,
        activation: activation_fn,
        activation_key: activation,
        error: error_fn,
        error_key: error,
        neuron_count: count,
        weights: Matrix::from(&vec![0.0]),
        weights_transpose: Matrix::from(&vec![0.0]),
    }
}

File: javascript/src/Layer.ts
import { ActivationFunction, ActivationFunctionKey } from "./Activation";
import { ErrorFunction, ErrorFunctionKey } from "./Error";
import { Matrix } from "./Matrix";

export type ForwardPassResult = {
  /**
   * Neuron sum of weights*inputs
   */
  sum: Matrix;
  /**
   * Sum passed through the activation function
   */
  activated: Matrix;
  /**
   * Error of the result.
   */
  error: number;
};

export class Layer {
  isOutput: boolean = false;

  neuronCount: number;

  weights: Matrix = Matrix.fromList([]);

  // Transposed weights array, where bias terms have been omitted.
  weightsTranspose: Matrix = Matrix.fromList([]);

  activation: ActivationFunction;

  error: ErrorFunction;

  constructor(
    activation: ActivationFunctionKey,
    neuronCount: number,
    error: ErrorFunctionKey
  ) {
    this.neuronCount = neuronCount;

    this.activation = new ActivationFunction(activation);
    this.error = new ErrorFunction(error);
  }

  /**
   * Calculates the neuron sum, activation and the error (on output layer) for a given input-output pair.
   */
  forwardPass = (input: Matrix, expected: Matrix): ForwardPassResult => {
    const sum = this.weights.multiply(input);
    const activated = this.activation.forward(sum, this.isOutput);

    return {
      sum,
      activated,
      error: this.isOutput ? this.error.loss(activated, expected) : -1,
    };
  };

  /**
   * Calculate the backwardpass for the given forward-pass-result and the next weights and deltas.
   */
  backwardPass = (
    result: ForwardPassResult,
    deltaNext: Matrix,
    layerNext: Layer
  ): Matrix => {
    const weights = layerNext.weightsTranspose;

    return weights
      .multiply(deltaNext)
      .hadamard(this.activation.backward(result.sum), true);
  };

  /**
   * Calculate the output delta. Can be extended to use hard-coded activation/error pairs for a bit of a performance boost.
   *  - Example of cross-entropy with softmax activation.
   */
  outputPass = (result: ForwardPassResult, expected: Matrix) => {
    if (this.activation.type === "softmax") {
      if (this.error.type !== "cross-entropy") {
        throw new Error("Can only use softmax with cross-entropy");
      }
      return this.activation.output(result.activated, expected);
    }

    return this.error
      .grad(result.activated, expected)
      .hadamard(this.activation.backward(result.sum), true);
  };

  /**
   * Set the weights of the layer. Also computes the transpose for later use.
   */
  updateWeights = (weights: Matrix) => {
    this.weights = weights;
    this.weightsTranspose = weights.omit(0).transpose();
  };

  /**
   * Initializes the layer.
   */
  initialize = (weights: Matrix, isOutput = false) => {
    this.isOutput = isOutput;
    this.updateWeights(weights);
  };
}
